### ISSUE TITLE
Pyright accepts illegal use of `Self` in class declaration

### ISSUE THREAD

**Issue by erictraut** [microsoft/pyright#6805]
[PEP 637](https://peps.python.org/pep-0673/#valid-locations-for-self) indicates that the following should generate an error, but pyright doesn't currently do so.

```python
T = TypeVar("T")

class Bar(Generic[T]):
    def bar(self) -> T: ...

# This should generate an error.
class Baz(Bar[Self]): ...  # Rejected
```

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.343, which I just published. It will also be included in a future release of pylance.
